src/libraries/System.Linq/src/System/Linq/Shuffle.SpeedOpt.cs

```diff
@@ -68,6 +69,204 @@ public override int GetCount(bool onlyIfCheap) =>
+                if (_source.TryGetNonEnumeratedCount(out int sourceCount) && sourceCount <= count)
+                {
+                    return base.Take(count);
+                }
+
+                // Otherwise, we either don't know how many elements are in the source, or we know it's more than count.
+                // Try to optimize by using reservoir sampling to get a random sample of count elements.
+                return new TakeShuffleIterator<TSource>(_source, count);
+            }
+        }
+
+        private sealed partial class TakeShuffleIterator<TSource> : Iterator<TSource>
+        {
+            private readonly IEnumerable<TSource> _source;
+            private readonly int _takeCount;
+            private List<TSource>? _buffer;
+
+            public TakeShuffleIterator(IEnumerable<TSource> source, int takeCount)
+            {
+                Debug.Assert(source is not null);
+                Debug.Assert(takeCount > 0);
+
+                _source = source;
+                _takeCount = takeCount;
+            }
+
+            private protected override Iterator<TSource> Clone() => new TakeShuffleIterator<TSource>(_source, _takeCount);
+
+            public override bool MoveNext()
+            {
+                int state = _state;
+
+            Initialized:
+                if (state > 1)
+                {
+                    List<TSource>? buffer = _buffer;
+                    Debug.Assert(buffer is not null);
+
+                    int i = state - 2;
+                    if (i < buffer.Count)
+                    {
+                        _current = buffer[i];
+                        _state++;
+                        return true;
+                    }
+                }
+                else if (state == 1)
+                {
+                    List<TSource>? buffer = SampleToList(_source, _takeCount);
+                    if (buffer is not null)
+                    {
+                        _buffer = buffer;
+                        _state = state = 2;
+                        goto Initialized;
+                    }
+                }
+
+                Dispose();
+                return false;
+            }
+
+            public override void Dispose()
+            {
+                _buffer = null;
+                base.Dispose();
+            }
+
+            /// <summary>Uses reservoir sampling to randomly select <paramref name="takeCount"/> elements from <paramref name="source"/>.</summary>
+            private static List<TSource>? SampleToList(IEnumerable<TSource> source, int takeCount)
+            {
+                List<TSource>? reservoir = null;
+
+                if (source is IList<TSource> list)
+                {
+                    int listCount = list.Count;
+                    Debug.Assert(listCount > takeCount, "Known listCount <= takeCount should have been handled by Iterator.Take override");
+
+                    reservoir = new(takeCount);
```

Comment:
It's so special if `takeCount` is `1` so that the list allocated here (even though the capacity has been set) might be removed in theory. The cost is a `bool` branch and more code lines.

I don't know how to judge the balance 🤣 . It seems that randomly picking a single element is more often than other count numbers.

Reply:
> It seems that randomly picking a single element is more often than other count numbers.

How so?

I do not think it's worth spending even more code optimizing that case until it's proven to really be necessary. It would non-trivially complicate things, all to avoid an extra allocation, when we're already allocating for the Shuffle enumerable and the Take enumerable and whatever the input to the Shuffle was.

Reply:
> until it's proven to really be necessary

Got it.